plugins/signals-backend/src/service/SignalManager.ts

```diff
@@ -132,35 +144,54 @@ export class SignalManager {
       return;
     }
 
-    const { channel, recipients, message } = eventPayload;
+    const { channel, recipients, message, permissions } = eventPayload;
     const jsonMessage = JSON.stringify({ channel, message });
     let users: string[] = [];
     if (recipients !== null) {
       users = Array.isArray(recipients) ? recipients : [recipients];
     }
 
     // Actual websocket message sending
-    this.connections.forEach(conn => {
+    for (const conn of this.connections.values()) {
       if (!conn.subscriptions.has(channel)) {
-        return;
+        continue;
       }
+
       // Sending to all users can be done with null
       if (
         recipients !== null &&
         !conn.ownershipEntityRefs.some((ref: string) => users.includes(ref))
       ) {
-        return;
+        continue;
+      }
+
+      if (permissions) {
+        // In case the message requires permissions but we can't authorize, skip sending
+        if (!this.permissions) {
+          continue;
+        }
+
+        const decisions = await this.permissions.authorize(permissions, {
```

Comment:
shouldn't this check be done in the plugin sending the event instead? 🤔
What's the advantage of passing the plain permissions, performing the check at `signals-backend` level?

Reply:
This is not possible as the signals can have many receivers. Of course, if you send the signal to only the current user, the permission check can be done in the sender end. An example for this:

1. DevTools wants to broadcast updates to info data (CPU/mem usage etc.) for all users at the `devtools` page
2. To see the info in the first place, the user must have `devToolsInfoReadPermission`
3. `devtools-backend` cannot see which users have subscribed to the signals 
4. Thus the permission must be checked during sending and not in the `devtools-backend`

Reply:
> This is not possible as the signals can have many receivers

ah I see, thank you for clarifying!
How does the subscription flow work? For instance, as a Backstage user, how can I subscribe to updates for the devtools' plugin?

Reply:
![signals](https://github.com/backstage/backstage/assets/1178319/081147af-79a7-492e-af88-6af4510994f4)

Maybe this 5-minute drawing is the easiest way to explain the functionality of the signals. Does it make sense?

Reply:
Or maybe even better, here's PR for devtools that explains it in code: https://github.com/backstage/backstage/pull/23194 :)

Reply:
thank you so much. It makes sense.

Do you have any suggestions about how to potentially deal with an expired client token here?

Reply:
It's a bit of a hard thing to do. One option would be to send the token via websocket channel instead in headers and refresh it on the client side when needed. I have to think about this a bit but I think it's in another PR. If you guys have ideas, they are more than welcome 🙏

Reply:
Briefly mentioned in #23194, but I wonder if it's better to authorize permissions when subscribing to a signal channel? Just to avoid hammering the permissions backend for small frequent signals. That could also come with the added benefit that an up-to-date user token can be included in that request, fixing the token expiration problem.

Since there's no concept of persistent channels with metadata or anything like that, the permission would simply have to be matched against permissions required when sending a signal later on. i.e. if a user subscribes to channel `example.a` with permission `example.read`, then the signal backend would allow signals through on that channel that either do not require any permission, or require `example.read`, signals with any other permission would be rejected.

I think that flow might have the additional benefit that we can set up a request/response flow for the signal subscription where the frontend can get feedback on whether the user has permission or not. Ofc you'd usually want to check that upfront in the frontend with a separate check, but with this the failure mode can be a bit better than just complete silence, even if we just forward permissions errors to the `ErrorApi`.

Reply:
The problem with this is exactly with the channels as the signal manager does not know which channel would require which kind of permission. IMHO I don't think it's a good idea to mix the channel name with permissions as it:

1. Gets really confusing to the signal sender as they have to match the permission name and the signal channel. This can lead to errors especially when refactoring things
2. Doesn't allow resource permission checks. For example, some plugins might want to send information in the same channel about different resources and with this architecture, it's not possible

For the performance issue you mentioned in the other PR, I think this should be taken care of in the `PermissionService` instead. Checking the same access rights multiple times for the same token should be very fast as it happens also in other places in the system. Another option would be to cache the authorization results per connection but I think that's not the right way to go either as it only applies to signals. 

Personally, I would like to keep this as light as possible in the frontend side, you can subscribe to anything but there's no guarantee that you will ever get data from that channel.